pkg/client/tar/tar.go

```diff
@@ -1,47 +1,93 @@
 package tar
 
-import "github.com/jhoonb/archivex"
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/pkg/errors"
+)
 
 const (
 	PathSeparator = "/"
 )
 
-type Writer interface {
-	AddFile(string, string) error
-	AddAll(string) error
-}
+func addAll(tw *tar.Writer, dir string, ignorePatterns []string) error {
+	return filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return errors.Wrap(err, "on walk call")
+		}
 
-type File interface {
-	Create(string) error
-	Close() error
-	Writer
-}
+		for _, ip := range ignorePatterns {
+			if matched, _ := filepath.Match(ip, info.Name()); matched {
+				if info.IsDir() {
+					return filepath.SkipDir
+				}
+				return nil
+			}
+		}
 
-type TarFile struct {
-	writer *archivex.TarFile
-}
+		if info.IsDir() {
+			return nil
+		}
 
-func (t *TarFile) Create(name string) error {
-	t.writer = new(archivex.TarFile)
-	return t.writer.Create(name)
-}
+		basePath := fmt.Sprintf("%s%c", filepath.Clean(dir), filepath.Separator)
+		name := strings.Replace(path, basePath, "", 1)
+		if runtime.GOOS == "windows" {
+			path = strings.Replace(path, string(filepath.Separator), PathSeparator, -1)
+			name = strings.Replace(name, string(filepath.Separator), PathSeparator, -1)
+		}
 
-func (t *TarFile) AddFile(path, filename string) error {
-	return t.writer.AddFileWithName(path, filename)
+		return addFile(tw, path, name, info)
+	})
 }
 
-func (t *TarFile) AddAll(path string) error {
-	return t.writer.AddAll(path, false)
-}
+func addFile(tw *tar.Writer, path, name string, info os.FileInfo) error {
+	if !info.Mode().IsRegular() {
+		return nil
+	}
+
+	header, err := tar.FileInfoHeader(info, "")
+	if err != nil {
+		return errors.Wrap(err, "failed to build tar header")
+	}
+	header.Name = name
+
+	if err := tw.WriteHeader(header); err != nil {
+		return errors.Wrap(err, "failed to write tar header")
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return errors.Wrap(err, "failed to open file")
+	}
+	defer file.Close()
 
-func (t *TarFile) Close() error {
-	return t.writer.Close()
+	_, err = io.Copy(tw, file)
+	return errors.Wrap(err, "failed to copy file contents to tarball")
```

Comment:
Always get an error? 
if not, what do you thinking about?:
```go
if _, err := io.Copy(tw, file); err != nil {
    return errors.Wrap(err, "failed to copy file contents to tarball")
}
return nil
```
I'm not sure if `errors.Wraps` returns `nil` for non errors args

Reply:
Yes, it does.